### PR TITLE
Fix memory leaking when using STB_IMAGE=1

### PR DESCRIPTION
**Pull request by caiiiycuk**

### CODE REVIEW

**Comment by kripken** [comment]
I don't understand the SDL change - it just moves some function declarations out of the try-catch? Is this just a cleanup, or does it fix something?

About the rest, @juj should look at the memoryprofiler parts.

**Comment by caiiiycuk** [comment]
> I don't understand the SDL change - it just moves some function declarations out of the try-catch? Is this just a cleanup, or does it fix something?

Yes I just move functions out of try/catch block. There is memory leak when this functions inside a try/catch block. My test reproduce this leaking in FF and Chrome. I am not a big master of js, but as I understand that problem is in unspecified behavior when functions declared using they definitions inside blocks. Without this fix, function bellow does not overrides cleanup function!
```js
    function addCleanup(func) {
        var old = cleanup;
        cleanup = function added_cleanup() {
          old();
          func();
        }
      }
```

This problem can be alternatively fixed with declaring function as var cleanup = function() ... inside try/catch block. Also I have jsfiddle with problem demonstration https://jsfiddle.net/jb7z901c/ 

--
From here http://stackoverflow.com/a/22173438/1213348:

Until ES2015, the spec didn't cover what a JavaScript engine should do if you put a function declaration inside a control structure like try, if, switch, while, etc., like this:
```js
if (someCondition) {
    function foo() {    // <===== HERE THERE
    }                   // <===== BE DRAGONS
}
```
And since they're processed before step-by-step code is run, it's tricky to know what to do when they're in a control structure.

Although doing this wasn't specified until ES2015, it was an allowable extension to support function declarations in blocks. Unfortunately (and inevitably), different engines did different things.

As of ES2015, the specification says what to do. In fact, it gives three separate things to do:

* If in loose mode not on a web browser, the JavaScript engine is supposed to do one thing
* If in loose mode on a web browser, the JavaScript engine is supposed to do something else
* If in strict mode (browser or not), the JavaScript engine is supposed to do yet another thing
The rules for the loose modes are tricky, but in strict mode, function declarations in blocks are easy: They're local to the block (they have block scope, which is also new in ES2015), and they're hoisted to the top of the block.

**Comment by kripken** [comment]
I see, thanks. Yeah, this is a serious bug. Let's fix this with `var cleanup = ` then, I think that's clearer and more consistent with other code.

**Comment by caiiiycuk** [comment]
Fixed, also I revert changes in memoryprofiler.js

File: tests/sdl_stb_image_cleanup.c [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+}
+
+int main() {
+  rename("screenshot.not", "screenshot.jpg");
+  for (int i = 0; i < 20; ++i) {
+    loadImage("screenshot.jpg");
+  }
+
+  int result = EM_ASM_INT_V({
+    return Math.trunc(emscriptenMemoryProfiler.totalMemoryAllocated / 1024 / 1024);
```
Comment (kripken) [inline tests/sdl_stb_image_cleanup.c]: to make sure I understand, the test is verifying that less than 1MB was allocated?

Comment (caiiiycuk) [inline tests/sdl_stb_image_cleanup.c]: Exactly

Comment (kripken) [inline tests/sdl_stb_image_cleanup.c]: Great, thanks.

**Pull request merged**